cosima_cookbook/querying.py

```diff
@@ -284,6 +293,14 @@ def _ncfiles_for_variable(
     if frequency is not None:
         q = q.filter(f.frequency == frequency)
 
+    # Attributes that are required to be unique to ensure disambiguation
+    for attr, val in attrs_unique.items():
+        # If default attribute present and not currently in filter
+        # add to attributes filter
+        if attr not in attrs:
+            if q.filter(v.ncvar_attrs.any(name=attr, value=val)).first():
+                attrs.update(attrs_unique)
+
```

Comment:
I guess this could be something like:

```suggestion
    # Attributes that are required to be unique to ensure disambiguation
    for attr, val in attrs_unique.items():
        q = q.filter(~v.ncvar_attrs.any((NCAttribute.name == attr) & (NCAttribute.value != val)))
```

which would tolerate the case of any of the unique attributes not existing. Although, as Andrew pointed out, it might cause trouble to have a unique attribute default that excludes results.

I haven't looked into too much detail, but we could do a query like the following to determine whether any attributes have non-unique values across a queried set (I guess it should be in the pipeline before we add the attribute constraints!)

```python
q_attr = (
    q.from_self()
    .join(NCAttribute, NCAttribute.ncvar_id == NCVar.id)
    .group_by(NCAttribute.name_id)
    .having(func.count(distinct(NCAttribute.value_id)) > 1)
)
s.query(q_attr.exists()).scalar()
```

Reply:
Thanks for the suggestions @angus-g. As you say, the logic in the suggested change will filter variables which have the same attribute but a different value. So in this case:
```python
attrs_unique={'cell_methods': 'area: time: mean'},
```
variables with `cell_methods = 'time: mean'` are filtered. 

The current logic will not filter these, but the check for unique values will flag the duplicated `cell_methods`:
```
UserWarning: Your query returns variables from 
files with different cell_methods: {'time: mean', 'time: mean_pow(02)'}. This could lead to unexpected behaviour!
Disambiguate by passing attrs={'cell_methods'=''} to getvar, specifying the desired attribute value.
```

So `attrs_unique` is performing two functions, providing a preferred default value if the attribute is present, but even if there is a different value for that attribute it will check for uniqueness. I like this, feels about the right level of intervention. So maybe this is about as good as it gets if that is the required functionality.

So then, should the check for unique attributes be done by the SQL logic you've suggested instead? When I ran it the query isn't returning any information about which attributes aren't unique. I guess it would need to be formulated  to select the attribute and the respective count?

Not sure it should be done before filtering. Isn't this designed to flag where filtering has failed to disambiguate?

I acknowledge I asked for feedback and have said "yeah nah", sorry about that, but thanks again for the suggestions.

Reply:
Okay, I understand, and yep my proposed change was slightly different to that logic. Makes sense! I think there may still be a bug in your implementation: `attrs` is updated with `attrs_unique` for _any_ name/value match, not the specific one that you're testing. Unless that is intentional?

> So then, should the check for unique attributes be done by the SQL logic you've suggested instead?

I guess the SQL method has the advantage that it can aggregate over all attributes, and all files, simultaneously. It's also pretty easy for it to check that the entire attribute set is consistent over the results, rather than just those specified in `attrs_unique`. Maybe that's a bit far-reaching though, it could be annoying to have something like a checksum or time info that's different over all files always triggering the warning. It would be pretty easy to tweak the query to select the offending attributes, if that's a path you want to explore. Otherwise, the loop over the returned objects seems fine.

> Not sure it should be done before filtering. Isn't this designed to flag where filtering has failed to disambiguate?

True, good point!

Reply:
> I think there may still be a bug in your implementation: `attrs` is updated with `attrs_unique` for _any_ name/value match, not the specific one that you're testing. Unless that is intentional?

No not intentional. Thanks for the heads-up.

Yeah I think warning about every inconsistent attribute is likely to be counter-productive, as false positives would just train people to ignore them. I'll leave as-is for the moment then. Tah.